### DISCUSSION TITLE
Service container: use the image built in the job

### DISCUSSION THREAD

**Discussion by SilleBille** [community/community#27147] [Actions]
<p>Hello everyone,</p>
<p>I’m new to Github actions and I’m experimenting with <a href="https://help.github.com/en/actions/building-actions/creating-a-docker-container-action" rel="nofollow noopener">Docker Container Action</a> and <a href="https://help.github.com/en/actions/configuring-and-managing-workflows/about-service-containers" rel="nofollow noopener">Service containers</a></p>
<p>In brief, my workflow looks like this:</p>
<ol>
<li>Build 2 different container images with required software packages</li>
<li>Run these container as 2 services</li>
<li>Run tests, which accesses both these service containers</li>
</ol>
<p>As per the <a href="https://help.github.com/en/actions/configuring-and-managing-workflows/about-service-containers#creating-service-containers" rel="nofollow noopener">official documentation</a>, these images are pulled from docker hub.</p>
<p>So, my questions are:</p>
<ol>
<li>Does GH actions support running service containers using the image created from the current (or previous) job, that runs as part of the commit/PR?</li>
<li>If <span class="hashtag">#1</span> is not feasible, the fallback approach I’m thinking of is to upload the image to <a href="http://quay.io" rel="nofollow noopener">quay.io</a> registry, rather than dockerhub. Can GH action upload container images to <a href="http://quay.io" rel="nofollow noopener">quay.io</a>?</li>
<li>Can service containers use image from <a href="http://quay.io" rel="nofollow noopener">quay.io</a>?</li>
</ol>
<p>Thanks,</p>
<p>Dinesh</p>

**Comment by amymariaparker2401** [comment]
<p>v2.0yyd57ah8ihbek6u5mdi</p>

**Comment by Yanjingzhu** [comment, marked as answer]
<blockquote>
<ol>
<li>Does GH actions support running service containers using the image created from the current (or previous) job, that runs as part of the commit/PR?</li>
<li>If <span class="hashtag">#1</span> is not feasible, the fallback approach I’m thinking of is to upload the image to <a href="http://quay.io" rel="nofollow noopener">quay.io</a> registry, rather than dockerhub. Can GH action upload container images to <a href="http://quay.io" rel="nofollow noopener">quay.io</a>?</li>
<li>Can service containers use image from <a href="http://quay.io" rel="nofollow noopener">quay.io</a>?</li>
</ol>
</blockquote>
<ol>
<li>
<p>The answer is no. </p>
</li>
<li>
<p>You could publish your image to <a href="http://quay.io" rel="nofollow noopener">quay.io</a> during GitHub workflow execution. You could run docker commands to login to <a href="http://quay.io" rel="nofollow noopener">quay.io</a> and then upload image to it. </p>
</li>
<li>
<p>If your <a href="http://quay.io" rel="nofollow noopener">quay.io</a> registry is a publlic one, it could be used in service container image key. If it is a private one, there is no way to provide credentials for the registry in service container section .  So it could not pull the image from <a href="http://quay.io" rel="nofollow noopener">quay.io</a>. </p>
</li>
</ol>

**Comment by SilleBille** [comment]
<p>Thanks for the response.</p>
<p>I can keep my quay repos public.</p>
<p>Do you or the concerned team have plans to implement <span class="hashtag">#1</span>? Or do you have feature request to host a temporary container registry within GH actions in your roadmap?</p>
<p>Also, I have a question unrelated to the original thread. Since the limitation provided by shared runners “might” not suffice my requirements to run multiple service containers, I was planning to use self-hosted runners. However, I was very worried about the <a href="https://help.github.com/en/actions/hosting-your-own-runners/about-self-hosted-runners#self-hosted-runner-security-with-public-repositories" rel="nofollow noopener">warning provided</a>. Do you have any plans to address these security concerns in your roadmap? My project is opensource and is prone for PRs from the community. If this question needs to be in a new thread, I can start one too :slight_smile:</p>
<p>–Dinesh</p>

**Comment by jcadavez** [comment]
<p>Friendly followup, are there plans to implement <span class="hashtag">#1</span>?</p>
<p>I’m facing the same issue except my images are private. For context, I’m running tests per each PR, this requires building images derived from the PR.</p>

**Comment by airtower-luna** [comment]
<aside class="quote no-group" data-username="jcadavez" data-post="5" data-topic="18198">
<div class="title">
<div class="quote-controls"></div>
 jcadavez:</div>
<blockquote>
<p>For context, I’m running tests per each PR, this requires building images derived from the PR.</p>
</blockquote>
</aside>
<p>How about running the images using Docker commands within the same job? I have <a href="https://github.com/airtower-luna/referer-mod/blob/e164091df63c8aa29c7b429cf126b43dbb2db17a/.github/workflows/selenium.yaml#L28-L35">a job</a> that builds and runs container images for tests using Docker Compose.</p>

**Comment by jcadavez** [comment]
<p>That’s a good suggestion and I have tried it. But, I need to be able to access each container’s feeds via <a href="http://localhost/" rel="noopener nofollow ugc">http://localhost</a>:port. And some of my images are on Amazon ECR, so I don’t think I can pull them with service-containers jobs (thus why I build them)</p>
<p>Unfortunately, the docs states that I have to use service-containers to access localhost.</p>

**Comment by airtower-luna** [comment]
<aside class="quote no-group" data-username="jcadavez" data-post="8" data-topic="18198">
<div class="title">
<div class="quote-controls"></div>
 jcadavez:</div>
<blockquote>
<p>Unfortunately, the docs states that I have to use service-containers to access localhost.</p>
</blockquote>
</aside>
<p>I don’t see it say that anywhere. You should be able to use <code>--publish</code> as usual when starting containers in your workflow.</p>
<p>You can <a href="https://docs.github.com/en/free-pro-team@latest/actions/reference/workflow-syntax-for-github-actions#jobsjob_idservicesservice_idcredentials">use credentials</a> when defining service containers, or <code>docker login</code> like on the command line, if authentication is the issue with Amazon ECR.</p>

**Comment by jcadavez** [comment]
<p>Sorry, my bad. It’s not in the docs, but I experienced it when I docker-compose w/o a service-container, and got <a href="https://stackoverflow.com/questions/62504968/github-workflow-econnrefused-127-0-0-1-error-when-connecting-to-docker-co" rel="noopener nofollow ugc">this issue</a></p>
<pre><code class="lang-auto">Error: connect ECONNREFUSED 127.0.0.1:8088
      at TCPConnectWrap.afterConnect [as oncomplete]
</code></pre>
<p>I can try Amazon ECR. My last resort is pulling the image created per the PR branch repo.</p>

**Comment by jcadavez** [comment]
<p>For better context, let’s say there’s four services: <strong>A</strong>, <strong>B</strong>, <strong>C</strong> and <strong>D</strong>.</p>
<p>They <code>depends-on</code> each other like below if in <em>docker-compose.yml</em>.</p>
<p><strong>A</strong> -&gt; <strong>B</strong> -&gt; <strong>C</strong> -&gt; <strong>D</strong>.</p>
<p>But, I’ll be committing a PR for <strong>C</strong>’s repo.</p>
<p>Each PR will trigger a GA Workflow that includes building <strong>C</strong> 's image with the others somehow per <a href="https://docs.github.com/en/free-pro-team@latest/actions/guides/about-service-containers" rel="noopener nofollow ugc">service-containers</a> method.</p>
<p>If I don’t, when I try accessing <strong>C</strong> 's localhost url, I’ll get the below error (<a href="https://stackoverflow.com/questions/62504968/github-workflow-econnrefused-127-0-0-1-error-when-connecting-to-docker-co" rel="noopener nofollow ugc">thread</a>).</p>
<pre><code class="lang-auto">Error: connect ECONNREFUSED 127.0.0.1:8088
      at TCPConnectWrap.afterConnect [as oncomplete]
</code></pre>

**Comment by airtower-luna** [comment]
<p>It’s hard to say what should or shouldn’t work without seeing your container setup, I’m afraid. One thing I can say is that Actions jobs set up their own network for <code>services</code> and job <code>container</code> (if any), so <em>mixing</em> those with Docker Compose or <code>docker run</code> is going to be a bit tricky. The <a href="https://docs.github.com/en/free-pro-team@latest/actions/reference/context-and-expression-syntax-for-github-actions#job-context"><code>job</code> context</a> contains some information on that network.</p>

**Comment by jcadavez** [comment]
<p>Thank you for your help so far Luna. I’ll look into your resource.</p>
<p>If this provides more context, I added a template of my <em>docker-compose.yml</em> below. Each of them except the db images have a repo. They depend on each other as follows</p>
<p><strong>C</strong> -&gt; <strong>B</strong> + <strong>A</strong> -&gt; <strong>db</strong></p>
<p>I start my GA workflow like below:</p>
<ol>
<li>Start a PR for <strong>C</strong> and it triggers the GA Workflow</li>
<li>Checkout each of the above repos (except db, that’s public)</li>
<li>Build each of their images</li>
<li><code>docker-compose up</code></li>
<li>Checkout my tests repo and run them against the <strong>C</strong>’s localhost url (<a href="http://localhost:3333" rel="noopener nofollow ugc">http://localhost:3333</a>).</li>
</ol>
<p>I keep getting the <code>EconnRefused</code> error. But, I want to get the feeds from <strong>C</strong> via its url. My team wants to avoid pulling the repos’ images, just build them.</p>
<pre><code class="lang-auto">version: "3"

services:
  db:
    image: mongo:latest
    container_name: db
  db-seed:
    image: mongo:latest
    container_name: db_seed
    links:
      - db
    volumes_from:
      - db
    command:
      /import.sh
  A:
    image: ecr-aws/A
    container_name: api
    ports:
        - 1111:1111
    env_file:
      - 'A/.env'
    depends_on: 
      - db
  B:
    image: ecr-aws/B
    container_name: B
    ports:
      - 2222:2222
    env_file:
      - 'B/.env'
    depends_on: 
      - db
  C:
    image: ecr-aws/C
    container_name: C
    ports:
      - 3333:3333
    env_file:
      - 'C/.env'
    depends_on:
      - A
</code></pre>

**Comment by airtower-luna** [comment]
<p>From the compose file connecting to localhost:3333 should work, assuming the service doesn’t fail to start during some other issue. I think any further analysis would need at least the workflow (to see if there’s any bug there), and ideally logs.</p>
